### PR TITLE
Adding workloads at risk dashboard

### PR DESCRIPTION
**Pull request by aburhan**
This dashboard is used to surface besteffort and burstable workloads running above requested resources (CPU and memory) with scale recommendations for deployments

### CODE REVIEW

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        },
+        "width": 2,
+        "xPos": 10,
+        "yPos": 0
+      },
+      {
+        "height": 2,
+        "widget": {
+          "text": {
+            "content": "**Performance Reliability Risk**: For **memory**: when workloads are running above requests, it has the same problem as best effort pods (can be killed at any time). For **CPU**: when workloads are running above requests, slowing down the workload\n [Set appropriate resource requests and limits](https://cloud.google.com/architecture/best-practices-for-running-cost-effective-kubernetes-applications-on-gke#set_appropriate_resource_requests_and_limits)",
```
Comment (yqlu) [inline dashboards/google-kubernetes-engine/gke-workloads-at-risk.json:123]: How about

Pods using more **memory** than requested constitute a reliability risk. Like best effort pods above, they can be stopped by the kubelet anytime. When workloads consume more **CPU** than requested, they can be throttled, slowing down the workload. [Set appropriate resource requests and limits](...)

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+            }
+          },
+          "title": "MEM under-provisioned"
```
Comment (yqlu) [inline dashboards/google-kubernetes-engine/gke-workloads-at-risk.json:170]: Let's be consistent and use Memory everywhere instead of MEM

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+            "dataSets": [
+              {
+                "timeSeriesQuery": {
+                  "timeSeriesQueryLanguage": "fetch k8s_container\n| metric 'kubernetes.io/container/memory/request_bytes'\n| filter resource.namespace_name !~ '(kube|istio|gatekeeper|gke|gmp)-system'\n| group_by 5m, [value_request_bytes_mean: mean(value.request_bytes)]\n| every 5m\n| group_by\n    [resource.project_id, resource.location, resource.cluster_name,\n     resource.namespace_name,\n     metadata.system.top_level_controller_name:\n       metadata.system_labels.top_level_controller_name,\n     metadata.system.top_level_controller_type:\n       metadata.system_labels.top_level_controller_type],\n    [value_request_bytes_mean_mean: mean(value_request_bytes_mean)]\n| filter val() == 0"
+                }
+              }
+            ],
+            "metricVisualization": "NUMBER"
+          },
+          "title": "Memory (reliability risk) - workloads without memory requested"
```
Comment (yqlu) [inline dashboards/google-kubernetes-engine/gke-workloads-at-risk.json:211]: Shorten this to "Workloads without memory requested"

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+            "dataSets": [
+              {
+                "timeSeriesQuery": {
+                  "timeSeriesQueryLanguage": "fetch k8s_container\n| metric 'kubernetes.io/container/cpu/request_cores'\n| filter resource.namespace_name !~ '(kube|istio|gatekeeper|gke|gmp)-system'\n| group_by 5m, [value_request_cores_mean: mean(value.request_cores)]\n| every 5m\n| group_by\n    [resource.project_id, resource.location, resource.cluster_name,\n     resource.namespace_name,\n     metadata.system.top_level_controller_name:\n       metadata.system_labels.top_level_controller_name,\n     metadata.system.top_level_controller_type:\n       metadata.system_labels.top_level_controller_type],\n    [value_request_bytes_mean_mean: mean(value_request_cores_mean)]\n| filter val() == 0"
+                }
+              }
+            ],
+            "metricVisualization": "NUMBER"
+          },
+          "title": "CPU (reliability risk) - workloads without CPU requested"
```
Comment (yqlu) [inline dashboards/google-kubernetes-engine/gke-workloads-at-risk.json:256]: Shorten this to "Workloads without CPU requested"

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        },
+        "width": 6,
+        "xPos": 6,
+        "yPos": 16
+      },
+      {
+        "height": 2,
+        "widget": {
+          "text": {
+            "content": "Recommended resource requests for CPU and memory requirements, ensuring efficient resource based on VPA recommendations.\n\n**Want to see workload recommendations at scale?**\n[Right-size your GKE workloads](https://cloud.google.com/blog/products/containers-kubernetes/optimize-gke-workloads-at-scale)",
```
Comment (yqlu) [inline dashboards/google-kubernetes-engine/gke-workloads-at-risk.json:354]: How about something like 

> Refer to the following recommended resource requests for CPU and memory requirements that ensuring efficient resource allocation based on VPA recommendations. They are calculated as 14 day peak recommendations.
> 
> **Want to see workload recommendations at scale?** [Right-size your GKE workloads](https://cloud.google.com/blog/products/containers-kubernetes/optimize-gke-workloads-at-scale)

and no title, if we've moved the title up into the collapsible group

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json (hunk 1 above)
Comment (aburhan) [reply dashboards/google-kubernetes-engine/gke-workloads-at-risk.json]: updated

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json (hunk 2 above)
Comment (aburhan) [reply dashboards/google-kubernetes-engine/gke-workloads-at-risk.json]: updated

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json (hunk 4 above)
Comment (aburhan) [reply dashboards/google-kubernetes-engine/gke-workloads-at-risk.json]: updated

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+            "dataSets": [
+              {
+                "timeSeriesQuery": {
+                  "timeSeriesQueryLanguage": "fetch k8s_scale\n| metric\n    'kubernetes.io/autoscaler/container/memory/per_replica_recommended_request_bytes'\n| group_by 14d,\n    [value_per_replica_recommended_request_bytes_mean:\n       max(value.per_replica_recommended_request_bytes)]\n| every 5m\n\n"
+                }
+              }
+            ],
+            "metricVisualization": "NUMBER"
+          },
+          "title": "MEM Recommended per replica request bytes [MAX]"
```
[the remark this answers is not in the document]
Comment (aburhan) [reply dashboards/google-kubernetes-engine/gke-workloads-at-risk.json]: updated

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json (hunk 5 above)
Comment (aburhan) [reply dashboards/google-kubernetes-engine/gke-workloads-at-risk.json]: updated

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+        "yPos": 10
+      },
+      {
+        "height": 4,
+        "widget": {
+          "timeSeriesTable": {
+            "dataSets": [
+              {
+                "timeSeriesQuery": {
+                  "timeSeriesQueryLanguage": "fetch k8s_container\n| metric 'kubernetes.io/container/memory/request_utilization'\n| filter\n    (resource.namespace_name !~ '(kube|istio|gatekeeper|gke|gmp)-system'\n     && metric.memory_type == 'non-evictable')\n| group_by 5m, [value_request_utilization_mean: mean(value.request_utilization)]\n| every 5m\n| group_by\n    [resource.project_id, resource.location, resource.cluster_name,\n     resource.namespace_name, metadata.system_labels.top_level_controller_name,\n     metadata.system_labels.top_level_controller_type],\n    [value_request_utilization_mean: mean(value_request_utilization_mean)]\n| filter val() >= 1\n| scale '%'"
```
[the remark this answers is not in the document]
Comment (aburhan) [reply dashboards/google-kubernetes-engine/gke-workloads-at-risk.json]: updated

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+                }
+              }
+            ],
+            "metricVisualization": "NUMBER"
+          },
+          "title": "Memory (reliability risk) - workloads running above requested for memory"
```
[the remark this answers is not in the document]
Comment (aburhan) [reply dashboards/google-kubernetes-engine/gke-workloads-at-risk.json]: updated

**Comment by aburhan** [comment]
> Thanks for contributing! Some minor formatting suggestions for the dashboard. Tagging in the alerting team to review the alert policies.

updated with formatting suggestions

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.json [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+        },
+        "width": 2,
+        "xPos": 2,
+        "yPos": 0
+      },
+      {
+        "height": 2,
+        "widget": {
+          "text": {
+            "content": "Not setting CPU or memory requests constitutes a reliability risk. When memory is unset, the kubelet can stop your pods anytime and mark them as failed. When CPU is unset, your Pods' CPU can be throttled to 0, causing your workloads to become unresponsive[Set appropriate resource requests and limits](https://cloud.google.com/architecture/best-practices-for-running-cost-effective-kubernetes-applications-on-gke#set_appropriate_resource_requests_and_limits)",
```
Comment (yqlu) [inline dashboards/google-kubernetes-engine/gke-workloads-at-risk.json:116]: Add a period and space between "to become unresponsive" and the "Set appropriate resource..." links.

**Review by yqlu — APPROVED** [review]
Just one more text formatting issue, then LGTM for the dashboard. 

Wait for an approval for the alert policies before merging please.

File: dashboards/google-kubernetes-engine/gke-workloads-at-risk.01.png
[the remark this answers is not in the document]
Comment (aburhan) [reply dashboards/google-kubernetes-engine/gke-workloads-at-risk.01.png:1]: updated the screenshot

**Review by yqlu — APPROVED** [review]
